cedar-policy/src/api/err.rs

```diff
@@ -517,45 +517,122 @@ impl From<cedar_policy_validator::ValidationWarning> for ValidationWarning {
     }
 }
 
+/// Error structs for the variants of `PolicySetError`
+pub mod policy_set_error_structs {
+    use super::Error;
+    use crate::PolicyId;
+    use miette::Diagnostic;
+
+    /// There was a duplicate [`PolicyId`] encountered in either the set of
+    /// templates or the set of policies.
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("duplicate template or policy id `{id}`")]
+    pub struct AlreadyDefined {
+        pub(crate) id: PolicyId,
+    }
+
+    /// Expected a static policy, but a template-linked policy was provided
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("expected a static policy, but a template-linked policy was provided")]
+    pub struct ExpectedStatic {}
+
+    /// Expected a template, but a static policy was provided.
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("expected a template, but a static policy was provided")]
+    pub struct ExpectedTemplate {}
+
+    /// Error when removing a static policy that doesn't exist
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("unable to remove static policy `{policy_id}` because it does not exist")]
+    pub struct PolicyNonexistentError {
+        pub(crate) policy_id: PolicyId,
+    }
+
+    /// Error when removing a static policy that doesn't exist
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("unable to remove template `{template_id}` because it does not exist")]
+    pub struct TemplateNonexistentError {
+        pub(crate) template_id: PolicyId,
+    }
+
+    /// Error when removing a template with active links
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("unable to remove policy template `{template_id}` because it has active links")]
+    pub struct RemoveTemplateWithActiveLinksError {
+        pub(crate) template_id: PolicyId,
+    }
+
+    /// Error when removing a template that is not a template
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("unable to remove policy template `{template_id}` because it is not a template")]
+    pub struct RemoveTemplateNotTemplateError {
+        pub(crate) template_id: PolicyId,
+    }
+
+    /// Error when unlinking a template
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
+    #[error("unable to unlink policy `{policy_id}` because it does not exist")]
+    pub struct LinkNonexistentError {
+        pub(crate) policy_id: PolicyId,
+    }
+
+    /// Error when removing a link that is not a link
+    #[derive(Debug, Diagnostic, PartialEq, Eq, Error, Clone)]
```

Comment:
Add these derives to `PolicySetError` also? Not much use otherwise.

Reply:
`serde_json::Error` doesn't derive these, so `PolicySetError` can't either. I was thinking they'd still be useful for testing, but since none of our tests use them and you can't construct the structs outside of this crate I think we can remove them.